backends/dpdk/dpdkArch.cpp

```diff
@@ -334,14 +334,69 @@ const IR::Node *ReplaceMetadataHeaderName::preorder(IR::Member *m) {
     return m;
 }
 
-const IR::Node *ReplaceMetadataHeaderName::preorder(IR::Parameter *p) {
-    if (auto type = p->type->to<IR::Type_Name>()) {
-        if (type->path->name == info->header_type) {
-            return new IR::Parameter(IR::ID("h"), p->direction, p->type);
-        } else if (type->path->name == info->local_metadata_type) {
-            return new IR::Parameter(IR::ID("m"), p->direction, p->type);
+/* This function replaces the header and metadata parameter names in control
+ * blocks with "h" and "m" respectively.
+ * It assumes that ConvertToDpdkArch pass has converted the control blocks to
+ * the following form for DPDK Architecture:
+ *         control ingressDeparser(packet_out bufffer, header hdr, metadata md);
+ *         control egressDeparser(packet_out bufffer, header hdr, metadata md);
+ *         control ingress(header hdr, metadata md);
+ *         control egress(header hdr, metadata md);
+*/
+const IR::Node *ReplaceMetadataHeaderName::preorder(IR::Type_Control *c) {
+    auto applyParams = new IR::ParameterList();
+    auto paramSize = c->applyParams->size();
+
+    if (paramSize == 2 || paramSize == 3) {
+        int header_index = 0;
+
+        /* For IngressDeparser and EgressDeparser, Header and metadata parameters are
+           at index 1 and 2 respectively. */
+        if (paramSize == 3) {
+            header_index = 1;
+            applyParams->push_back(c->getApplyParameters()->getParameter(0));
         }
+
+        auto header = c->applyParams->parameters.at(header_index);
+        auto local_metadata = c->applyParams->parameters.at(header_index + 1);
+        header = new IR::Parameter(IR::ID("h"), header->direction, header->type);
+        local_metadata = new IR::Parameter(IR::ID("m"), local_metadata->direction,
+                                           local_metadata->type);
+        applyParams->push_back(header);
+        applyParams->push_back(local_metadata);
+
+        return new IR::Type_Control(c->name, c->annotations, c->typeParameters,
+                                    applyParams);
     }
+
+    return c;
+}
+
+/* This function replaces the header and metadata parameter names in parser
+ * blocks with "h" and "m" respectively.
+ * It assumes that ConvertToDpdkArch pass has converted Ingress/Egress
+ * parser blocks to the following form for DPDK Architecture:
+ *         parser ingressParser(packet_in bufffer, header hdr, metadata md);
+ *         parser egressParser(packet_in bufffer, header hdr, metadata md);
+*/
+const IR::Node *ReplaceMetadataHeaderName::preorder(IR::Type_Parser *p) {
+    auto applyParams = new IR::ParameterList();
+    auto paramSize = p->applyParams->size();
+
+    if (paramSize == 3) {
```

Comment:
Wouldn't you want to give an error if this is not true?
Here is how the bmv2 backend does error handling: https://github.com/p4lang/p4c/blob/5fa90dae7ebba31e5aa71bc7580edacb02ea502a/backends/bmv2/simple_switch/simpleSwitch.cpp#L43

Reply:
Updated. I think at this point of compilation, it should be a bug if we do not have the right number of parameters. It is based on the assumption that a previous pass has converted the program to this expected format.

Reply:
The rule of thumb I have followed is that users should never see a compiler bug message, no matter what input they supply to the compiler. To stress test this you can try invoking the dpdk compiler without using psa as a model. Just try compiling a regular v1model.p4 program with this backend.

Reply:
I got your point. I will update the BUG to a modelError as you suggested. But, that also raises a doubt about the usage of BUG for error handling. In what scenarios would we want to emit a message using BUG? I see a lot of places where it is used in dpdk backend.

I also tried compiling the v1model.p4 program with p4c-dpdk without --arch psa option, we currently emit a model error for PSA followed by a Compiler Bug.

[--Werror=Target model error] error: Package package V1Switch has no parameter named 'ingress'Are you using an up-to-date 'psa.p4'?
         DeparserImpl()) main;
In file: ../ir/ir.cpp:230
Compiler Bug: No parameter named ingress

Shouldn't we exit at the model error itself?

Reply:
You can choose where you want to exit.
Usually we finish the current pass and after each pass check for errors and exit.
This allows us to report multiple errors in one pass.
Another solution is to make the compiler bug conditional on no errors:
`BUG_CHECK(condition && errorCount() == 0, ...);`